Index: xen-4.7.1/xen/arch/x86/efi/efi-tboot.h
===================================================================
--- /dev/null
+++ xen-4.7.1/xen/arch/x86/efi/efi-tboot.h
@@ -0,0 +1,6 @@
+#ifndef __EFITBOOT_H__
+#define __EFITBOOT_H__
+
+/* Information passed to Xen at launch by TBOOT */
+
+#endif /* __EFITBOOT_H__ */
Index: xen-4.7.1/xen/arch/x86/tboot.c
===================================================================
--- xen-4.7.1.orig/xen/arch/x86/tboot.c
+++ xen-4.7.1/xen/arch/x86/tboot.c
@@ -18,6 +18,9 @@
 static unsigned long __initdata opt_tboot_pa;
 integer_param("tboot", opt_tboot_pa);
 
+/* Global pointer to shared data passed by EFI boot code. */
+void *g_efi_tboot_shared;
+
 /* Global pointer to shared data; NULL means no measured launch. */
 tboot_shared_t *g_tboot_shared;
 
